### PR TITLE
Add _CAPn and _TRIG environemnt vars

### PR DESCRIPTION
**Pull request by leiserfg**
- Add _CAPn and _TRIG environment variables
- Auto generate docs

### CODE REVIEW

**Comment by L3MON4D3** [comment]
Ah, nice :+1:

TBH, I'm not a big fan of calling the variables `_CAP` and `_TRIG`, IMO we should just take the breaking change (affecting probably nobody) for nicer variable names.

File: lua/luasnip/util/environ.lua [hunk 1]
```diff
@@ -39,14 +39,14 @@ function Environ.is_table(var_fullname)
...
-function Environ:new(pos, o)
+function Environ:new(meta, o)
 	o = o or {}
 	setmetatable(o, self)
 
 	for ns_name, ns in pairs(namespaces) do
 		local eager_vars = {}
 		if ns.init then
-			eager_vars = ns.init(pos)
+			eager_vars = ns.init(meta)
```
Comment (L3MON4D3) [inline lua/luasnip/util/environ.lua:49]: Mhmm, what do you think of changing the table passed to `init` to
```lua
{
  pos = {row,col},
  trigger = ...
  captures = {...}
}
```
?
We could set `meta[1], meta[2]` (btw what's the mnemonic here? metadata?) to `row`, `col` to stay backward compatible

**Comment by leiserfg** [comment]
> Ah, nice +1
> 
> TBH, I'm not a big fan of calling the variables `_CAP` and `_TRIG`, IMO we should just take the breaking change (affecting probably nobody) for nicer variable names.

this one is up to you

File: lua/luasnip/util/environ.lua (hunk 1 above)
Comment (leiserfg) [reply lua/luasnip/util/environ.lua:49]: > btw what's the mnemonic here? metadata? 
It's either that or a reference to the effects of  Meta over the environment :interrobang: 




> We could set `meta[1], meta[2]`

Do you think then I should do that and keep `meta[1] & meta[2]` just as an undocumented non-breaking change protection?

Comment (L3MON4D3) [reply lua/luasnip/util/environ.lua:49]: > > btw what's the mnemonic here? metadata?
> 
> It's either that or a reference to the effects of Meta over the environment :interrobang:

:laughing:
Mhmm, what `data` is this the `metadata` to though?
(Just to be clear, I'm not worried about the name of the variable itself, more about what we call it in the documentation. I think `meta` is a bit confusing (+negatively connotated :P), just `info` maybe?)

> > We could set `meta[1], meta[2]`
> 
> Do you think then I should do that and keep `meta[1] & meta[2]` just as an undocumented non-breaking change protection?

Yes, undocumented :+1::+1:

**Comment by L3MON4D3** [comment]
Hmmmm yeah, a namespace for our variables is a good idea, a breaking change for each variable we add would be annoying.

TBH, I'd prefer the longer versions `LS_CAPTUREn`, `LS_TRIGGER`, feels more in line with other variables.

I'm thinking; should we add older variables we introduced (`SELECT_RAW`, `SELECT_DEDENT`) to this namespace too?
Upside, it would be more coherent, downside, breaking change (and this one would actually affect ppl)
=> I think we shouldn't.

**Comment by leiserfg** [comment]
> I'm thinking; should we add older variables we introduced (`SELECT_RAW`, `SELECT_DEDENT`) to this namespace too?

I will do both, so in the doc we can just talk about `LS` but we don't break old users' configs.

**Comment by leiserfg** [comment]
> > > I'm thinking; should we add older variables we introduced (`SELECT_RAW`, `SELECT_DEDENT`) to this namespace too?
> > 
> > 
> > I will do both, so in the doc we can just talk about `LS` but we don't break old users' configs.
> 
> True, that sounds like the best solution, as long as it's not too complex to maintain (but doesn't seem to be) +1

In case you didn't notice, it's done already.

**Comment by L3MON4D3** [comment]
Yup :+1:
Mhmmm I think it's okay to just leave lambdas implementation of `l.CAPTUREn` alone, it works, it'll continue to work.

Would you add tests for the new stuff passed to `init`, and `LS_`-variables?

Apart from that, this is pretty clean, nice job :+1::+1:

**Comment by L3MON4D3** [comment]
Nice, looks good :+1:
Wanna squash them, 1 commit?

File: DOC.md [hunk 2]
```diff
@@ -1436,9 +1436,11 @@ You can also add your own variables by using the `ls.env_namespace(name, opts)`
     Is a function that receives a string and returns a value for the var with that name
     or a table from var name to a value
     (in this case, if the value is a function it will be executed  lazily once per snippet expansion).
-    * `init`: `fn(pos: pair[int])->map[string, EnvVal]` Takes the 0-based position of the cursor and returns
+    * `init`: `fn(info: pair[int])->map[string, EnvVal]`  Returns
```
Comment (bew) [inline DOC.md:1439]: note: `info` is not really a `pair[int]` anymore 'officially'

File: tests/unit/environ_spec.lua [hunk 3]
```diff
@@ -180,4 +187,26 @@ describe("luasnip.util.environ", function()
...
+						var_name,
+						test
+					)
+				)
+			)
+		end)
+	end
+
+	check_builtin("LS_TRIGGER", [[function(r) return r == "trigg" end]])
+	check_builtin("LS_CAPTURE1", [[function(r) return r == "one" end]])
```
Comment (bew) [inline tests/unit/environ_spec.lua:211]: if we're going for long name, for capture group, it could be named `LS_CAPTURE_1` ?

Comment (leiserfg) [reply tests/unit/environ_spec.lua:211]: [«Is mir egal»](https://www.youtube.com/watch?v=YEYim54pJ00) @L3MON4D3  should decide.

Comment (L3MON4D3) [reply tests/unit/environ_spec.lua:211]: :D
I agree, it's easier to read :+1:

File: DOC.md [hunk 4]
```diff
@@ -1399,7 +1399,7 @@ If `jsregexp` is not available, transformation are replaced by a simple copy.
 # VARIABLES
 
 All `TM_something`-variables are supported with two additions:
-`SELECT_RAW` and `SELECT_DEDENT`. These were introduced because
+`LS_SELECT_RAW` and `LS_SELECT_DEDENT`. These were introduced because
```
Comment (L3MON4D3) [inline DOC.md:1402]: Ah, could you add the new vars here too?
Best make it a list for readability

Comment (leiserfg) [reply DOC.md:1402]: I think we should (not in this PR) list the TM_ ones too.

Comment (leiserfg) [reply DOC.md:1402]: A link is ok for the online doc, but for the .txt I think it's better to have it in place :thinking:

**Comment by L3MON4D3** [comment]
(I think test failure in d69e20b is just due to the testing-framework, seen it happen a few times)

File: doc/luasnip.txt [hunk 5]
```diff
@@ -1345,6 +1345,9 @@ and `LS_SELECT_DEDENT`. These were introduced because `TM_SELECTED_TEXT` is
 designed to be compatible with vscodes’ behavior, which can be
 counterintuitive when the snippet can be expanded at places other than the
 point where selection started (or when doing transformations on selected text).
+Besides those we also provide `LS_TRIGGER` which contains the trigger of the
+snippet, and `LS_CAPTURE_n` (where n is a positive integer) that contains the
+n-th capture of the when using a regex as `trig` in the snippet definition.
```
Comment (bew) [inline doc/luasnip.txt:1350]:
```suggestion
n-th capture when using a regex with capture groups as `trig` in the snippet definition.
```

Comment (leiserfg) [reply doc/luasnip.txt]: So I can just use your review directly :smile:

**Comment by leiserfg** [comment]
If everyone is happy now I will squash this.

**Comment by leiserfg** [comment]
Squashed and re-named

**Comment by L3MON4D3** [comment]
Ah, could you add a notice of the `LS`-breaking change to the commitmessage?

**Comment by leiserfg** [comment]
> Ah, could you add a notice of the `LS`-breaking change to the commitmessage?

I will but if someone is using it it will appear anyway because of the asserts in environ.

File: DOC.md (hunk 4 above)
Comment (L3MON4D3) [reply DOC.md:1402]: Hmm yes, good point... I don't think `panvimdoc` allows .txt-specific sections though.
OTOH, just making a big list should be okay for the markdown as well, then we could even mention details of our implementation.

**Comment by L3MON4D3** [comment]
True, but don't wanna be the bad guy if someone updates their config and it suddenly throws a big scary _red_ (:P) error

**Comment by leiserfg** [comment]
> True, but don't wanna be the bad guy if someone updates their config and it suddenly throws a big scary _red_ (:P) error

You mean "again" :smile:

**Comment by L3MON4D3** [comment]
Fine work @leiserfg :ok_hand:

**Pull request merged**